contrib/colab/DC-GAN.ipynb

```diff
@@ -0,0 +1,554 @@
+        "\n",
+        "Therefore the training loop in this notebook will entail:\n",
+        "1. Update the generator ($G$) to maximize the probability of the discriminator making the incorrect choice on generated data:\n",
+        "$$\\underset{G}{\\text{maximize}}\\;  \\mathbb{E}_{z \\sim p(z)}\\left[\\log D(G(z))\\right]$$\n",
+        "2. Update the discriminator ($D$), to maximize the probability of the discriminator making the correct choice on real and generated data:\n",
+        "$$\\underset{D}{\\text{maximize}}\\; \\mathbb{E}_{x \\sim p_\\text{data}}\\left[\\log D(x)\\right] + \\mathbb{E}_{z \\sim p(z)}\\left[\\log \\left(1-D(G(z))\\right)\\right]$$"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "o6r3OlA_QhxP",
+        "colab": {}
+      },
+      "source": [
+        "import torch\n",
+        "from torch import nn, optim\n",
+        "from torchvision import transforms, datasets\n",
+        "from torch.optim import Adam\n",
+        "import torch.nn.functional as F\n",
+        "\n",
+        "import torch_xla\n",
+        "import torch_xla.core.xla_model as xm\n",
+        "import torch_xla.debug.metrics as met\n",
+        "import torch_xla.distributed.parallel_loader as pl\n",
+        "import torch_xla.distributed.xla_multiprocessing as xmp\n",
+        "import torch_xla.utils.utils as xu"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "colab_type": "text",
+        "id": "AYxAwFtDQhxR"
+      },
+      "source": [
+        "# Setting up the Global Flags\n",
+        "\n",
+        "In the current setup, Discriminator network was chosen to be a smaller capacity than generator. Even with similar capacity networks, generator update path is deeper than discriminator. Therefore uneven learning rates chosen here seems to yield a better convergence. \n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "fDStxFFGQhxR",
+        "colab": {}
+      },
+      "source": [
+        "# Define Parameters\n",
+        "FLAGS = {}\n",
+        "FLAGS['datadir'] = \"/tmp/mnist\"\n",
+        "FLAGS['batch_size'] = 128\n",
+        "FLAGS['num_workers'] = 4\n",
+        "FLAGS['gen_learning_rate'] = 0.005\n",
+        "FLAGS['disc_learning_rate'] = 0.001\n",
+        "FLAGS['num_epochs'] = 21\n",
+        "FLAGS['num_cores'] = 8  "
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "R40fkk90QhxT",
+        "colab": {}
+      },
+      "source": [
+        "%matplotlib inline\n",
```

Comment:
do we need this? we're saving to file and then loading it. Is it needed when plotting the loaded image?

Reply:
You are right, we can remove it.